Спринт 15 Домашнее задание

AlexanderKorytin opened the pull request:
Все поправил

AlexanderKorytin committed:
Спринт 14
1. изменена анимация Scale в MediaActivity на кнопку playPause
2. в MediaActivity нажатие на кнопку playPause переписано через setOnTouchListener для более интересной работы анимаций

AlexanderKorytin committed:
Спринт 15
1. Создана структура под Чистую Архитектуру
2. В MediaActivity работа с библиотекой Glide реализована через интерфейс ImageLoader и его реализацию ImageLoaderGlide
3. Создан интерфейс для работы с сетью NetworkClient
4. Написана реализация NetworkClient -> RetrofitNetworkClient
5. Изменен TracksResponse
6. создана модель TrackDto для раборы в слое data
7. создан TrackRequest
8. создан
8. создан Responce() - родитель для всех ответов сервера
9. Создан интерфейс для работы с сетью TracksRepository
10. Написана реализация TracksRepository -> TracksRepositoryImpl

AlexanderKorytin committed:
Спринт 15
1. написан интерфейс TracksInteractor
2. создана реализация интерфейса TracksInteractor -> TracksRepositoryImpl
3. До следующего сприната, пока в FindActivity в getMusic сделан map из TrackDto в Track
4. в domain создан dataclass ClickedTrack для использования в медиаплеере
5. написан interface GetClickedTrackFromData и его реализация GetClickedTrackFromGson для перевода Gson полученого из intent в медиаплеере в ClickedTrack
6. написан interface SetViewVisibilityUseCase и его реализация SetViewVisibilityClearButton

AlexanderKorytin committed:
Спринт 15
Реализации интерфейсов ImageLoader и GetClickedTrackFromGson перенесены в слой data

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/Find/FindActivity.kt:
```diff
@@ -1,4 +1,4 @@
-package com.example.playlistmaker
```
пакет `Find` нужно с маленькой буквы назвать

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/dto/Responce.kt:
```diff
@@ -0,0 +1,5 @@
+package com.example.playlistmaker.data.dto
+
+open class Responce() {
```
```suggestion
open class Response() {
```
Опечатка

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/NetworkClient.kt:
```diff
@@ -0,0 +1,7 @@
+import com.example.playlistmaker.data.dto.Responce
+
+interface NetworkClient {
+    fun doRequest(dto: Any): Responce
```
```suggestion
    fun searchTracks(request: TrackRequest): Responce
```
Для каждого нового запроса здесь будет создаваться новая функция, а не использоваться универсальная `doRequest`

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/dto/GetClickedTrackFromGson.kt:
```diff
@@ -0,0 +1,13 @@
+import com.example.playlistmaker.domain.models.ClickedTrack
+import com.google.gson.Gson
+
+class GetClickedTrackFromGson(): GetClickedTrackFromData {
```
```suggestion
class GetClickedTrackFromGsonUseCase(): GetClickedTrackFromDataUseCase {
```
UseCase'ы принято называть таким образом, чтобы в конце шла строка "UseCase", например:
`GetDataUseCase`
`SaveDataUseCase`

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/dto/ImageLoaderGlide.kt:
```diff
@@ -0,0 +1,20 @@
+
+class ImageLoaderGlide: ImageLoader {
+
+    override fun loadImage(url: String, placeholder: Int, imageView: ImageView, round: Int) {
```
```suggestion
    override fun loadImage(url: String, placeholder: Int, imageView: ImageView, @Px cornerRadius: Int) {
```
Так будет понятнее за что именно отвечает 4й параметр и в каком формате используется

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/network/RetrofitNetworkClient.kt:
```diff
@@ -0,0 +1,26 @@
+        .build()
+    private val iTunesService = retrofit.create(ITunesApi::class.java)
+    override fun doRequest(dto: Any): Responce {
+        if (dto is TrackRequest){
```
Есть еще замечания по код стилям, в Android студии можно применить автоформатирование, нажав ПКМ по нужному классу и выбрав `Reformat code`, или нажать сочетание клавиш из соответствующего пункта в меню.

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/domain/api/GetClickedTrackFromData.kt:
```diff
@@ -0,0 +1,7 @@
+import com.example.playlistmaker.domain.models.ClickedTrack
+
+interface GetClickedTrackFromData {
+    fun execute(track: Any?): ClickedTrack
```
Параметр у UseCase'a не должен быть `Any?`. Нужно четко определить с чем работает юзкейс

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/domain/api/SetViewVisibility.kt:
```diff
@@ -0,0 +1,5 @@
+package com.example.playlistmaker.domain.api
+
+interface SetViewVisibility {
+    fun execute(s: Any?): Boolean
```
1. Нужно переименовать класс и назвать его UseCase'ом
2. Параметр функции `execute` должен быть четко определён

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/domain/api/TracksInteractor.kt:
```diff
@@ -0,0 +1,11 @@
+interface TracksInteractor {
+    fun getMusic(expression: String, consumer: TracksConsumer)
+
+    interface TracksConsumer {
```
Можно не писать свой такой интерфейс, в JDK уже есть интерфейс Consumer<T>

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/domain/impl/SetViewVisibilityClearButton.kt:
```diff
@@ -0,0 +1,9 @@
+
+class SetViewVisibilityClearButton: SetViewVisibility {
+    override fun execute(s: Any?): Boolean {
+        return (s as CharSequence).isNotEmpty()
```
Раз юзкейс работает только с `CharSequence` в качестве параметра, то так и стоит написать:
`override fun execute(s: CharSequence): Boolean`

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/domain/impl/TracksInteractorImpl.kt:
```diff
@@ -0,0 +1,14 @@
+class TracksInteractorImpl(private  val repositiry: TracksRepository): TracksInteractor {
+
+    override fun getMusic(term: String, consumer: TracksInteractor.TracksConsumer) {
+        val t = Thread{
```
Ты хотел вынести вызов функции репозитория в отдельный поток, но использовать `Thread` не стоит. К тому же делать это стоит не в интеракторе. Пока я бы убрал это отсюда и сделал просто вызов `repository.getMusic(term)`. Асинхронная работа будет в следующих спринтах

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/presentetion/mediaplayer/MediaPlayerPresenter.kt:
```diff
@@ -0,0 +1,4 @@
+package com.example.playlistmaker.presentetion.mediaplayer
+
+class MediaPlayerPresenter {
```
Лишний класс

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/presentetion/mediaplayer/MediaplayerView.kt:
```diff
@@ -0,0 +1,4 @@
+package com.example.playlistmaker.presentetion.mediaplayer
+
+class MediaplayerView {
```
Лишний класс

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/Find/FindActivity.kt:
```diff
@@ -186,7 +203,21 @@ class FindActivity : AppCompatActivity() {
                                     setVisibilitySearchСompleted()
                                     trackList.clear()
-                                    trackList.addAll(response.body()?.results!!)
+                                    trackList.addAll(response.body()?.results!!.map {
```
Не нужно использовать `!!`, т.к. это приведёт к крешу приложения. Лучше обработать ситуацию, когда здесь будет null - например показать сообщение пользователю, что по его запросу ничего не найдено

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/ui/Find/FindViewHolder.kt:
```diff
@@ -14,19 +16,21 @@ class FindViewHolder(private val parentView: View) : RecyclerView.ViewHolder(par
     private val trackTime: TextView = parentView.findViewById(R.id.trackTime)
     private val radiusIconTrackDp = 2.0f
-    private val correction = 0.5f
+    val imageLoaderGlide = ImageLoaderGlide()
```
Лучше передать в конструкторе интерфейс `ImageLoader`

AlexanderKorytin committed:
Спринт 15
поправил недочеты по коментариям PullRequest

AlexanderKorytin committed:
Спринт 15
Исправлены критические замечания

AlexanderKorytin committed:
Спринт 15
реализация getMusic(text: String, adapter: FindAdapter) в FindActivity переделана с использование чистой архитектуры

AlexanderKorytin committed:
Спринт 15
- бота с историей поиска вынесена в Репозиторий
- написан интерфейс SearchHistory и его реализация SearchHistoryImpl
- Создан класс SearchHistoryUseCase для работы с историей поиска из FindActivity

AlexanderKorytin committed:
Спринт 15
выполнен remname функции ImageLoaderUseCase (loadd -> execute) и класса SearchHistoryUseCase -> SearchHistoryInteractor

AlexanderKorytin committed:
Спринт 15
приведен в соответсвии с логикой textSearch -> text в блоке if getMusic

AlexanderKorytin committed:
Спринт 15
- внесены изменения отображения видимости View в bindingFindActivity.placeholderButton.setOnClickListenerWithViber
- внесено условие проверки флага flagIOException в getMusic для случая непустого листа, с целью нормального отображения плэсхолдера "Проблемы со связью" после удачного поискового запроса

AlexanderKorytin committed:
Спринт 15
- все Activity помещены в пакет presentetion\ui

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/data/network/RetrofitNetworkClient.kt:
```diff
@@ -1,26 +1,24 @@
-        } else {
-            return Responce().apply { resultCode = 400 }
+    override fun searchTracks(request: TrackRequest): Response {
+        val response = iTunesService.searchTracks(request.term).execute()
```
Я бы посоветовал сейчас не делать запросы с помощью `execute`, а использовать `enqueue` и прокидывать в функции колбэки

ArturNurtdinov commented on app/src/main/java/com/example/playlistmaker/presentetion/ui/find/FindActivity.kt:
```diff
@@ -187,61 +184,34 @@ class FindActivity : AppCompatActivity() {
-                })
+        if (text.isNotEmpty()) {
+            adapter.notifyDataSetChanged()
+            val t = Thread {
```
Чтоб вот здесь не приходилось создавать целый поток, я бы советовал работать через колбэки и запросом в сеть с помощью `enqueue`

AlexanderKorytin committed:
- создан EditTextInteractor для работы с EditText. Работа с Android SDK в domain?
